# Analysis of Neighborhoods in Toronto, Canada
## Part 3 - Neighborhoods Clustering, Plotting, and Analysis

Marnilo C

In [1]:
# Import libraries
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

# for easier handling or URLs
import requests

# for parsing HTML
from bs4 import BeautifulSoup

print('Libraries imported.')

Libraries imported.


### Create the dataframe with Postal Codes, Boroughs, Neighborhoods, Latitude, and Longitude

This task uses three dataframes for clarity.
1. df1 is for Postal Codes, Boroughs, and Neighborhoods. The data is obtained by scraping https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M.
1. df2 is for Postal Codes, Latitude, and Longitude. The data was imported from the CSV file at http://cocl.us/Geospatial_data.
1. Toronto_df is is for Postal Codes, Boroughs, Neighborhoods, Latitude, and Longitude.

In [4]:
# Create the dataframe df1 for Postal Codes, Boroughs, and Neighborhoods

# Scrape the data
url1 = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
html_data = requests.get(url1).text

soup = BeautifulSoup(html_data,'lxml')
#print(soup.prettify())

html_table = soup.find('table',{'class':'wikitable sortable'})

# Create the dataframe.
def parse_html_table(table):
            n_columns = 0
            n_rows=0
            column_names = []
    
            # Find number of rows and columns
            # we also find the column titles if we can
            for row in table.find_all('tr'):
                
                # Determine the number of rows in the table
                td_tags = row.find_all('td')
                if len(td_tags) > 0:
                    n_rows+=1
                    if n_columns == 0:
                        # Set the number of columns for our table
                        n_columns = len(td_tags)
                        
                # Handle column names if we find them
                th_tags = row.find_all('th') 
                if len(th_tags) > 0 and len(column_names) == 0:
                    for th in th_tags:
                        column_names.append(th.get_text())
    
            # Safeguard on Column Titles
            if len(column_names) > 0 and len(column_names) != n_columns:
                raise Exception("Column titles do not match the number of columns")
    
            columns = column_names if len(column_names) > 0 else range(0,n_columns)
            df = pd.DataFrame(columns = columns,
                              index= range(0,n_rows))
            row_marker = 0
            for row in table.find_all('tr'):
                column_marker = 0
                columns = row.find_all('td')
                for column in columns:
                    df.iat[row_marker,column_marker] = column.get_text().strip()
                    column_marker += 1
                if len(columns) > 0:
                    row_marker += 1
                    
            # Convert to float if possible
            for col in df:
                try:
                    df[col] = df[col].astype(float)
                except ValueError:
                    pass
            
            return df

df1 = parse_html_table(html_table)

# Clean the dataframe
df1.rename(columns={'Postcode': 'PostalCode', 'Neighbourhood\n': 'Neighborhood'}, inplace=True)
df1.drop(df1.loc[df1.Borough == 'Not assigned'].index, inplace=True)
df1['Neighborhood'] = df1['Neighborhood'].str.strip()
df1.loc[df1['Neighborhood'] == 'Not assigned', 'Neighborhood'] = df1['Borough']
df1 = df1.groupby('PostalCode').agg({'Borough':'first',
                               'Neighborhood': ', '.join}).reset_index()

# Create the dataframe df2 for Postal Codes and Geodata.

# This section is needed only if the CSV data needs to be download.
# !wget -q -O 'toronto_geodata.csv' http://cocl.us/Geospatial_data
# print('CSV data downloaded.')

df2 = pd.read_csv('toronto_geodata.csv', header=0)
df2.rename(columns={'Postal Code': 'PostalCode'}, inplace=True)

# Create new dataframe with boroughs, neighborhoods, and geodata
Toronto_df = pd.merge(df1, df2, on='PostalCode', how='inner')

print('Dataframe Toronto_df shape:', Toronto_df.shape)
print('Dataframe Toronto_df has {} boroughs and {} Neighborhoods.'.format(
        len(Toronto_df['Borough'].unique()),
        Toronto_df.shape[0]
    )
)
Toronto_df.head()

Dataframe Toronto_df shape: (103, 5)
Dataframe Toronto_df has 11 boroughs and 103 Neighborhoods.


PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Rouge, Malvern  43.806686   
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

## Analyze Neighborhoods in Toronto

### Prepare the map of Toronto Neighborhoods

#### Use geopy library to get the latitude and longitude values of Toronto.
In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>to_explorer</em>, as shown below.

In [5]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Toronto are 43.653963, -79.387207.


#### Create the map of Toronto with markers for Neighborhoods

In [6]:
# create map of New York using latitude and longitude values
Toronto_map = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, Neighborhood in zip(Toronto_df['Latitude'], Toronto_df['Longitude'], Toronto_df['Borough'], Toronto_df['Neighborhood']):
    label = '{}, {}'.format(Neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(Toronto_map)  
    
Toronto_map

In [16]:
# OPTIONAL: limit the analysis to Neighborhoods in one borough.

### Explore Neighborhoods

#### Define FourSquare API credentials and version

In [7]:
CLIENT_ID = 'JLNSZ0ZGFBI12HAKMFQL0QDGNFKHXMLP52VKCVSW4ZAOKHU2' # your Foursquare ID
CLIENT_SECRET = '4FLNSFIKFOTYK2UTX53G442NAX11BHA4FZOFIY4HG1A1NDJ2' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: JLNSZ0ZGFBI12HAKMFQL0QDGNFKHXMLP52VKCVSW4ZAOKHU2
CLIENT_SECRET:4FLNSFIKFOTYK2UTX53G442NAX11BHA4FZOFIY4HG1A1NDJ2


#### Define function to obtain venues

In [10]:
def getNearbyVenues(names, latitudes, longitudes, radius=500, LIMIT=100):

    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Create a new dataframe for venues

In [11]:
Toronto_venues = getNearbyVenues(names=Toronto_df['Neighborhood'],
                                   latitudes=Toronto_df['Latitude'],
                                   longitudes=Toronto_df['Longitude']
                                  )

Rouge, Malvern
Highland Creek, Rouge Hill, Port Union
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park, Ionview, Kennedy Park
Clairlea, Golden Mile, Oakridge
Cliffcrest, Cliffside, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Scarborough Town Centre, Wexford Heights
Maryvale, Wexford
Agincourt
Clarks Corners, Sullivan, Tam O'Shanter
Agincourt North, L'Amoreaux East, Milliken, Steeles East
L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
Silver Hills, York Mills
Newtonbrook, Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park, Don Mills South
Bathurst Manor, Downsview North, Wilson Heights
Northwood Park, York University
CFB Toronto, Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Woodbine Gardens, Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West, 

#### Check the resulting dataframe

In [12]:
print(Toronto_venues.shape)
Toronto_venues.head()

(2241, 7)


Neighborhood  Neighborhood Latitude  \
0                          Rouge, Malvern              43.806686   
1  Highland Creek, Rouge Hill, Port Union              43.784535   
2       Guildwood, Morningside, West Hill              43.763573   
3       Guildwood, Morningside, West Hill              43.763573   
4       Guildwood, Morningside, West Hill              43.763573   

   Neighborhood Longitude                            Venue  Venue Latitude  \
0              -79.194353                          Wendy's       43.807448   
1              -79.160497            Royal Canadian Legion       43.782533   
2              -79.188711  Swiss Chalet Rotisserie & Grill       43.767697   
3              -79.188711                G & G Electronics       43.765309   
4              -79.188711                       Marina Spa       43.766000   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.163085                   Bar  
2       -79.189914           Pizza Place  
3       -79.191537     Electronics Store  
4       -79.191000                   Spa

#### Check how many venues were returned for each neighborhood

In [13]:
Toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Agincourt                                                               4   
Agincourt North, L'Amoreaux East, Milliken, Ste...                      2   
Albion Gardens, Beaumond Heights, Humbergate, J...                     11   
Alderwood, Long Branch                                                  9   
Bathurst Manor, Downsview North, Wilson Heights                        18   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      26   
Berczy Park                                                            55   
Birch Cliff, Cliffside West                                             5   
Bloordale Gardens, Eringate, Markland Wood, Old...                      7   
Brockton, Exhibition Place, Parkdale Village                           19   
Business Reply Mail Processing Centre 969 Eastern                      18   
CFB Toronto, Downsview East                                             3   
CN Tower, Bathurst Quay, Island airport, Harbou...                     14   
Cabbagetown, St. James Town                                            42   
Caledonia-Fairbanks                                                     6   
Canada Post Gateway Processing Centre                                  10   
Cedarbrae                                                               7   
Central Bay Street                                                     88   
Chinatown, Grange Park, Kensington Market                             100   
Christie                                                               16   
Church and Wellesley                                                   88   
Clairlea, Golden Mile, Oakridge                                         9   
Clarks Corners, Sullivan, Tam O'Shanter                                10   
Cliffcrest, Cliffside, Scarborough Village West                         2   
Cloverdale, Islington, Martin Grove, Princess G...                      1   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             35   
Davisville North                                                        8   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                     14   
Del Ray, Keelesdale, Mount Dennis, Silverthorn                          6   
Design Exchange, Toronto Dominion Centre                              100   
Don Mills North                                                         6   
Dorset Park, Scarborough Town Centre, Wexford H...                      7   
Dovercourt Village, Dufferin                                           19   
Downsview Central                                                       3   
Downsview Northwest                                                     4   
Downsview West                                                          6   
Downsview, North Park, Upwood Park                                      4   
East Birchmount Park, Ionview, Kennedy Park                             4   
East Toronto                                                            4   
Emery, Humberlea                                                        2   
Fairview, Henry Farm, Oriole                                           64   
First Canadian Place, Underground city                                100   
Flemingdon Park, Don Mills South                                       21   
Forest Hill North, Forest Hill West                                     4   
Glencairn                                                               4   
Guildwood, Morningside, West Hill                                       8   
Harbord, University of Toronto                                         35   
Harbourfront East, Toronto Islands, Union Statio

#### Find out how many unique categories can be curated from all the returned venues

In [16]:
print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 274 uniques categories.


#### Check the number of neighborhoods

In [17]:
print('There are {} uniques neighborhoods.'.format(len(Toronto_venues['Neighborhood'].unique())))

There are 99 uniques neighborhoods.


### Analyze Each Neighborhood

In [18]:
# one hot encoding
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Toronto_onehot['Neighborhood'] = Toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]

Toronto_onehot.head()

Yoga Studio  Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0            0                  0               0                  0        0   
1            0                  0               0                  0        0   
2            0                  0               0                  0        0   
3            0                  0               0                  0        0   
4            0                  0               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   Auto Garage  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  \
0            0              0          0           0           0       0   
1            0              0          0           0           0       0   
2            0              0          0           0           0       0   
3            0              0          0           0           0       0   
4            0              0          0           0           0       0   

   Bank  Bar  Baseball Field  Baseball Stadium  Basketball Court  \
0     0    0               0                 0                 0   
1     0    1               0                 0                 0   
2     0    0               0                 0                 0   
3     0    0               0                 0                 0   
4     0    0               0                 0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bike Shop  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                   0          0       0              0          0         0   
1                   0          0       0              0          0         0   
2                   0          0       0              0          0         0   
3                   0          0       0              0          0         0   
4                   0          0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                     0               0        0            0   
1                     0               0        0            0   
2                     0               0        0            0   
3                     0               0        0            0   
4                     0               0 

In [23]:
# Examine the dataframe shape
Toronto_onehot.shape

(2241, 274)

#### Group rows by neighborhood and by the mean of the frequency of occurrence of each category

In [19]:
Toronto_grouped = Toronto_onehot.groupby('Neighborhood').mean().reset_index()
Toronto_grouped

Neighborhood  Yoga Studio  \
0                            Adelaide, King, Richmond     0.000000   
1                                           Agincourt     0.000000   
2   Agincourt North, L'Amoreaux East, Milliken, St...     0.000000   
3   Albion Gardens, Beaumond Heights, Humbergate, ...     0.000000   
4                              Alderwood, Long Branch     0.000000   
5     Bathurst Manor, Downsview North, Wilson Heights     0.000000   
6                                     Bayview Village     0.000000   
7                   Bedford Park, Lawrence Manor East     0.000000   
8                                         Berczy Park     0.000000   
9                         Birch Cliff, Cliffside West     0.000000   
10  Bloordale Gardens, Eringate, Markland Wood, Ol...     0.000000   
11       Brockton, Exhibition Place, Parkdale Village     0.000000   
12  Business Reply Mail Processing Centre 969 Eastern     0.055556   
13                        CFB Toronto, Downsview East     0.000000   
14  CN Tower, Bathurst Quay, Island airport, Harbo...     0.000000   
15                        Cabbagetown, St. James Town     0.000000   
16                                Caledonia-Fairbanks     0.000000   
17              Canada Post Gateway Processing Centre     0.000000   
18                                          Cedarbrae     0.000000   
19                                 Central Bay Street     0.011364   
20          Chinatown, Grange Park, Kensington Market     0.000000   
21                                           Christie     0.000000   
22                               Church and Wellesley     0.011364   
23                    Clairlea, Golden Mile, Oakridge     0.000000   
24            Clarks Corners, Sullivan, Tam O'Shanter     0.000000   
25    Cliffcrest, Cliffside, Scarborough Village West     0.000000   
26  Cloverdale, Islington, Martin Grove, Princess ...     0.000000   
27                     Commerce Court, Victoria Hotel     0.000000   
28                                         Davisville     0.000000   
29                                   Davisville North     0.000000   
30  Deer Park, Forest Hill SE, Rathnelly, South Hi...     0.000000   
31     Del Ray, Keelesdale, Mount Dennis, Silverthorn     0.000000   
32           Design Exchange, Toronto Dominion Centre     0.000000   
33                                    Don Mills North     0.000000   
34  Dorset Park, Scarborough Town Centre, Wexford ...     0.000000   
35                       Dovercourt Village, Dufferin     0.000000   
36                                  Downsview Central     0.000000   
37                                Downsview Northwest     0.000000   
38                                     Downsview West     0.000000   
39                 Downsview, North Park, Upwood Park     0.000000   
40        East Birchmount Park, Ionview, Kennedy Park     0.000000   
41                                       East Toronto     0.000000   
42                                   Emery, Humberlea     0.000000   
43                       Fairview, Henry Farm, Oriole     0.000000   
44             First Canadian Place, Underground city     0.000000   
45                   Flemingdon Park, Don Mills South     0.000000   
46                Forest Hill North, Forest Hill West     0.000000   
47                                          Glencairn     0.000000   
48                  Guildwood, Morningside, West Hill     0.000000   
49                     Harbord, University of Toronto     0.028571   
50  Harbourfront East, Toronto Islands, Union Station     0.000000   
51                          Harbourfront, Regent Park     0.022222   
52                      High Park, The Junction South     0.000000   
53             Highland Creek, Rouge Hill, Port Union     0.000000   
54                                  Hillcrest Village     0.000000   
55       Humber Bay Shores, Mimico South, New Toronto     0.000000   
56  Humber Bay, King's Mill Park, Kingsway Park So.

#### Check the shape of the new dataframe

In [20]:
Toronto_grouped.shape

(99, 274)

#### Print each neighborhood along with the top 5 most common venues

In [21]:
num_top_venues = 5

for hood in Toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = Toronto_grouped[Toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
             venue  freq
0      Coffee Shop  0.06
1             Café  0.05
2  Thai Restaurant  0.04
3       Steakhouse  0.04
4              Bar  0.04


----Agincourt----
            venue  freq
0          Lounge  0.25
1  Sandwich Place  0.25
2  Breakfast Spot  0.25
3    Skating Rink  0.25
4   Metro Station  0.00


----Agincourt North, L'Amoreaux East, Milliken, Steeles East----
                             venue  freq
0                       Playground   0.5
1                             Park   0.5
2                      Yoga Studio   0.0
3                      Men's Store   0.0
4  Molecular Gastronomy Restaurant   0.0


----Albion Gardens, Beaumond Heights, Humbergate, Jamestown, Mount Olive, Silverstone, South Steeles, Thistletown----
                 venue  freq
0        Grocery Store  0.18
1          Pizza Place  0.09
2  Japanese Restaurant  0.09
3          Coffee Shop  0.09
4       Discount Store  0.09


----Alderwood, Long Branch----
            v

#### Encode the information into a *pandas* dataframe

In [22]:
# Define function to to sort the venues in descending order.
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [26]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Toronto_grouped['Neighborhood']

for ind in np.arange(Toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

print('Dataframe neighborhoods_venues_sorted shape:', neighborhoods_venues_sorted.shape)
neighborhoods_venues_sorted.head()

Dataframe neighborhoods_venues_sorted shape: (99, 11)


Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                          Agincourt        Breakfast Spot   
2  Agincourt North, L'Amoreaux East, Milliken, St...                  Park   
3  Albion Gardens, Beaumond Heights, Humbergate, ...         Grocery Store   
4                             Alderwood, Long Branch           Pizza Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café            Steakhouse                   Bar   
1                Lounge          Skating Rink        Sandwich Place   
2            Playground         Women's Store      Doner Restaurant   
3           Coffee Shop        Discount Store              Pharmacy   
4           Coffee Shop                   Gym          Skating Rink   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0       Thai Restaurant   American Restaurant                Bakery   
1            Donut Shop                 Diner        Discount Store   
2      Department Store          Dessert Shop    Dim Sum Restaurant   
3           Pizza Place            Beer Store   Fried Chicken Joint   
4              Pharmacy                   Pub        Sandwich Place   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0          Burger Joint                 Hotel         Cosmetics Shop  
1               Dog Run      Doner Restaurant          Women's Store  
2                 Diner        Discount Store                Dog Run  
3        Sandwich Place  Fast Food Restaurant    Japanese Restaurant  
4                  Pool                 Diner          Deli / Bodega

## Cluster the Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [27]:
# set number of clusters
kclusters = 5

Toronto_grouped_clustering = Toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([3, 3, 0, 1, 1, 3, 3, 3, 3, 3], dtype=int32)

### Create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [31]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Toronto_merged = Toronto_df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
Toronto_merged = Toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood', how = 'inner')

print('Dataframe Toronto_merged shape:', Toronto_merged.shape)
Toronto_merged.head() # check the last columns!

Dataframe Toronto_merged shape: (99, 16)


PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Rouge, Malvern  43.806686   
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.194353               3  Fast Food Restaurant         Women's Store   
1 -79.160497               3                   Bar             Drugstore   
2 -79.188711               1     Electronics Store        Medical Center   
3 -79.216917               3           Coffee Shop     Korean Restaurant   
4 -79.239476               3    Athletics & Sports      Hakka Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0            Donut Shop          Dessert Shop    Dim Sum Restaurant   
1    Dim Sum Restaurant                 Diner        Discount Store   
2    Mexican Restaurant                   Spa   Rental Car Location   
3         Women's Store             Drugstore    Dim Sum Restaurant   
4                Bakery       Thai Restaurant  Caribbean Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                 Diner        Discount Store               Dog Run   
1               Dog Run      Doner Restaurant            Donut Shop   
2           Pizza Place          Intersection        Breakfast Spot   
3                 Diner        Discount Store               Dog Run   
4                  Bank   Fried Chicken Joint               Dog Run   

  9th Most Common Venue 10th Most Common Venue  
0      Doner Restaurant              Drugstore  
1         Women's Store                   Farm  
2    Dim Sum Restaurant                  Diner  
3      Doner Restaurant             Donut Shop  
4    Dim Sum Restaurant                  Diner

### Visualize the clusters on the map of Toronto

In [32]:
# create map
Toronto_clusters_map = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged['Latitude'], Toronto_merged['Longitude'], Toronto_merged['Neighborhood'], Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(Toronto_clusters_map)
       
Toronto_clusters_map

## Examine the clusters

### Cluster 1: Playgrounds

In [33]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 0, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
5       Scarborough               0            Playground   
14      Scarborough               0                  Park   
48  Central Toronto               0                   Gym   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
5          Women's Store            Donut Shop          Dessert Shop   
14            Playground         Women's Store      Doner Restaurant   
48            Playground            Donut Shop          Dessert Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
5     Dim Sum Restaurant                 Diner        Discount Store   
14      Department Store          Dessert Shop    Dim Sum Restaurant   
48    Dim Sum Restaurant                 Diner        Discount Store   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
5                Dog Run      Doner Restaurant              Drugstore  
14                 Diner        Discount Store                Dog Run  
48               Dog Run      Doner Restaurant              Drugstore

### Cluster 2: Pizza Places

In [34]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 1, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels  1st Most Common Venue 2nd Most Common Venue  \
2    Scarborough               1      Electronics Store        Medical Center   
13   Scarborough               1            Pizza Place    Chinese Restaurant   
15   Scarborough               1   Fast Food Restaurant    Chinese Restaurant   
24    North York               1            Coffee Shop           Pizza Place   
34    North York               1  Portuguese Restaurant           Coffee Shop   
35     East York               1   Fast Food Restaurant           Pizza Place   
72    North York               1            Pizza Place                Bakery   
81          York               1          Grocery Store           Pizza Place   
89     Etobicoke               1            Pizza Place           Coffee Shop   
95     Etobicoke               1      Convenience Store            Beer Store   
96    North York               1            Pizza Place   Empanada Restaurant   
99     Etobicoke               1           Intersection           Coffee Shop   
101    Etobicoke               1          Grocery Store           Coffee Shop   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
2      Mexican Restaurant                   Spa   Rental Car Location   
13    Fried Chicken Joint              Pharmacy  Fast Food Restaurant   
15               Gym Pool        Breakfast Spot        Discount Store   
24         Discount Store              Pharmacy         Grocery Store   
34            Pizza Place          Hockey Arena            Donut Shop   
35           Intersection                  Bank    Rock Climbing Spot   
72                    Pub   Japanese Restaurant         Women's Store   
81               Bus Line  Caribbean Restaurant         Women's Store   
89                    Gym          Skating Rink              Pharmacy   
95           Liquor Store              Pharmacy           Pizza Place   
96          Women's Store      Doner Restaurant          Dessert Shop   
99            Pizza Place        Sandwich Place    Chinese Restaurant   
101        Discount Store              Pharmacy           Pizza Place   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
2             Pizza Place          Intersection        Breakfast Spot   
13           Noodle House                  Bank    Italian Restaurant   
15         Cosmetics Shop              Pharmacy           Pizza Place   
24                Airport      Department Store           Event Space   
34           Dessert Shop    Dim Sum Restaurant                 Diner   
35     Athletics & Sports             Gastropub                  Café   
72       Doner Restaurant    Dim Sum Restaurant                 Diner   
81                Dog Run          Dessert Shop    Dim Sum Restaurant   
89                    Pub        Sandwich Place                  Pool   
95                   Café        Shopping Plaza            Donut Shop   
96     Dim Sum Restaurant                 Diner        Discount Store   
99          Women's Store      Doner Restaurant    Dim Sum Restaurant   
101            Beer Store   Fried Chicken Joint        Sandwich Place   

    9th Most Common Venue 10th Most Common Venue  
2      Dim Sum Restaurant                  Diner  
13        Thai Restaurant         Discount Store  
15         Sandwich Place            Coffee Shop  
24   Ethiopian Restaurant    Empanada Restaurant  
34         Discount Store                Dog Run  
35               Pharmacy              Pet Store  
72         Discount Store                Dog Run  
81                  Diner         Discount Store  
89                  Diner          Deli / Bodega  
95                  Diner         Discount Store  
96                Dog Run              Drugstore  
99                  Diner         Discount Store  
101  Fast Food Restaurant    Japanese Restaurant

### Cluster 3: Parks

In [35]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 2, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
23         North York               2                  Park   
25         North York               2              Bus Stop   
30         North York               2                  Park   
40          East York               2                  Park   
44    Central Toronto               2                  Park   
50   Downtown Toronto               2                  Park   
73               York               2                  Park   
74               York               2                  Park   
79         North York               2                  Park   
90          Etobicoke               2                  Park   
100         Etobicoke               2                  Park   

    2nd Most Common Venue       3rd Most Common Venue 4th Most Common Venue  \
23      Convenience Store                        Bank             Drugstore   
25                   Park        Fast Food Restaurant     Food & Drink Shop   
30      Other Repair Shop                     Airport         Women's Store   
40      Convenience Store                 Coffee Shop         Women's Store   
44               Bus Line                 Swim School      Doner Restaurant   
50                  Trail                  Playground      Department Store   
73                  Trail                       Field          Hockey Arena   
74   Fast Food Restaurant                    Pharmacy                Market   
79       Basketball Court  Construction & Landscaping                Bakery   
90                   Pool                       River    College Rec Center   
100           Pizza Place                    Bus Line      Doner Restaurant   

    5th Most Common Venue   6th Most Common Venue 7th Most Common Venue  \
23                  Diner          Discount Store               Dog Run   
25       Doner Restaurant            Dessert Shop    Dim Sum Restaurant   
30             Donut Shop            Dessert Shop    Dim Sum Restaurant   
40     Dim Sum Restaurant                   Diner        Discount Store   
44     Dim Sum Restaurant                   Diner        Discount Store   
50           Dessert Shop      Dim Sum Restaurant                 Diner   
73           Dessert Shop      Dim Sum Restaurant                 Diner   
74          Women's Store  Gluten-free Restaurant             Gift Shop   
79              Drugstore                   Diner        Discount Store   
90        College Stadium     Empanada Restaurant     Electronics Store   
100          Dessert Shop      Dim Sum Restaurant                 Diner   

           8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
23              Doner Restaurant            Donut Shop          Women's Store  
25                         Diner        Discount Store                Dog Run  
30                         Diner        Discount Store                Dog Run  
40                       Dog Run      Doner Restaurant             Donut Shop  
44                       Dog Run            Donut Shop       Department Store  
50                Discount Store               Dog Run       Doner Restaurant  
73                Discount Store               Dog Run       Doner Restaurant  
74   Eastern European Restaurant          Gourmet Shop    Dumpling Restaurant  
79                       Dog Run      Doner Restaurant             Donut Shop  
90   Eastern European Restaurant   Dumpling Restaurant              Drugstore  
100               Discount Store               Dog Run             Donut Shop

### Cluster 4: Coffee Shops and Cafes

In [36]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 3, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
0         Scarborough               3       Fast Food Restaurant   
1         Scarborough               3                        Bar   
3         Scarborough               3                Coffee Shop   
4         Scarborough               3         Athletics & Sports   
6         Scarborough               3             Discount Store   
7         Scarborough               3                   Bus Line   
8         Scarborough               3                      Motel   
9         Scarborough               3                       Farm   
10        Scarborough               3          Indian Restaurant   
11        Scarborough               3  Middle Eastern Restaurant   
12        Scarborough               3             Breakfast Spot   
17         North York               3                Golf Course   
18         North York               3             Clothing Store   
19         North York               3         Chinese Restaurant   
22         North York               3                 Restaurant   
26         North York               3                       Café   
27         North York               3           Asian Restaurant   
28         North York               3                Coffee Shop   
29         North York               3         Falafel Restaurant   
31         North York               3              Grocery Store   
32         North York               3             Baseball Field   
33         North York               3         Athletics & Sports   
36          East York               3               Skating Rink   
37       East Toronto               3              Grocery Store   
38          East York               3     Furniture / Home Store   
39          East York               3          Indian Restaurant   
41       East Toronto               3           Greek Restaurant   
42       East Toronto               3             Sandwich Place   
43       East Toronto               3                       Café   
45    Central Toronto               3              Grocery Store   
46    Central Toronto               3             Clothing Store   
47    Central Toronto               3                Pizza Place   
49    Central Toronto               3                Coffee Shop   
51   Downtown Toronto               3                Coffee Shop   
52   Downtown Toronto               3        Japanese Restaurant   
53   Downtown Toronto               3                Coffee Shop   
54   Downtown Toronto               3                Coffee Shop   
55   Downtown Toronto               3                Coffee Shop   
56   Downtown Toronto               3                Coffee Shop   
57   Downtown Toronto               3                Coffee Shop   
58   Downtown Toronto               3                Coffee Shop   
59   Downtown Toronto               3                Coffee Shop   
60   Downtown Toronto               3                Coffee Shop   
61   Downtown Toronto               3                Coffee Shop   
62         North York               3                  Juice Bar   
63    Central Toronto               3                     Garden   
64    Central Toronto               3                      Trail   
65    Central Toronto               3                       Café   
66   Downtown Toronto               3                       Café   
67   Downtown Toronto               3                       Café   
68   Downtown Toronto               3            Airport Service   
69   Downtown Toronto               3                Coffee Shop   
70   Downtown Toronto               3                Coffee Shop   
71         North York               3             Clothing Store   
75   Downtown Toronto               3              Grocery Store   
76       West Toronto               3                Supermarket   
77       West Toronto               3                        Bar   
78       West Toronto               3                Coffee Shop   
80

### Cluster 5: Cafeteria

In [37]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 4, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
20  North York               4             Cafeteria            Donut Shop   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
20          Dessert Shop    Dim Sum Restaurant                 Diner   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
20        Discount Store               Dog Run      Doner Restaurant   

   9th Most Common Venue 10th Most Common Venue  
20         Women's Store       Department Store

## Comments
1. I decided to use all the neighborhoods in Toronto instead of a limited set (for example, all Neighborhoods with the word "Toronto").

## Observations
1. The members of each cluster are distributed across the map. The clustering analysis did not directly translate to geographically distinct clusters on the map.
2. The number of neighborhoods in the dataframe was reduced from 103 (Toronto_df) to 99 (Toronto_venues). I assume that there might be some neighborhoods which do not have data in Foursquare.